<a href="https://colab.research.google.com/github/InsightfulSantosh/AWS_Sagemaker-flight-prices-prediction/blob/main/Data_Cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

###**Import Important Libriary**

In [1]:
import pandas as pd
import numpy as np

###**1.Load & read the Dataset**

In [2]:
flight= pd.read_csv('/content/flight_price.csv')

In [3]:
flight.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

In [4]:
flight.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


### **2. Preliminary Analysis**


####**2.1 check the data types**

In [5]:
flight.dtypes

Airline            object
Date_of_Journey    object
Source             object
Destination        object
Route              object
Dep_Time           object
Arrival_Time       object
Duration           object
Total_Stops        object
Additional_Info    object
Price               int64
dtype: object

####**2.2 check for missing values**

In [6]:
(
    pd.DataFrame({
        'Missing Count': flight.isna().sum(),
        'Missing Percent': (flight.isna().sum() / flight.shape[0]) * 100
    })
)


Missing Count  Missing Percent
Airline                      0         0.000000
Date_of_Journey              0         0.000000
Source                       0         0.000000
Destination                  0         0.000000
Route                        1         0.009361
Dep_Time                     0         0.000000
Arrival_Time                 0         0.000000
Duration                     0         0.000000
Total_Stops                  1         0.009361
Additional_Info              0         0.000000
Price                        0         0.000000

####**2.3 Check for duplicates**

In [7]:
(
    flight.
    duplicated().
    sum()
)

220

###**Observation**
- 1. Need to change the datatype of `Date_of_Journey`,`Dep_Time` & `Arrival_Time` to datetime.
- 2. The type of `Duration` and `Total_Stops` is mixed. It should be numeric & Categorical type respectively
- 3. There are 220 duplicates rows
- 4. There is only 1 missing value in each `Route` & `Total_Stops` .

### **3.  Detailed Analysis**

- 3.1 **Airline**

In [8]:
flight.Airline

0             IndiGo
1          Air India
2        Jet Airways
3             IndiGo
4             IndiGo
            ...     
10678       Air Asia
10679      Air India
10680    Jet Airways
10681        Vistara
10682      Air India
Name: Airline, Length: 10683, dtype: object

In [9]:
flight.Airline.unique()

array(['IndiGo', 'Air India', 'Jet Airways', 'SpiceJet',
       'Multiple carriers', 'GoAir', 'Vistara', 'Air Asia',
       'Vistara Premium economy', 'Jet Airways Business',
       'Multiple carriers Premium economy', 'Trujet'], dtype=object)

In [10]:
(
 flight.
 Airline.str.replace(' Premium economy',"").
         str.replace(" Business","").
         str.strip().
         str.title().
 unique()

)

array(['Indigo', 'Air India', 'Jet Airways', 'Spicejet',
       'Multiple Carriers', 'Goair', 'Vistara', 'Air Asia', 'Trujet'],
      dtype=object)

- Some of the entries have inconsistent/inaccurate values

- 3.2 **Date_of_Journey**

In [11]:
flight.Date_of_Journey

0        24/03/2019
1         1/05/2019
2         9/06/2019
3        12/05/2019
4        01/03/2019
            ...    
10678     9/04/2019
10679    27/04/2019
10680    27/04/2019
10681    01/03/2019
10682     9/05/2019
Name: Date_of_Journey, Length: 10683, dtype: object

In [12]:
flight.Date_of_Journey.unique()

array(['24/03/2019', '1/05/2019', '9/06/2019', '12/05/2019', '01/03/2019',
       '24/06/2019', '12/03/2019', '27/05/2019', '1/06/2019',
       '18/04/2019', '9/05/2019', '24/04/2019', '3/03/2019', '15/04/2019',
       '12/06/2019', '6/03/2019', '21/03/2019', '3/04/2019', '6/05/2019',
       '15/05/2019', '18/06/2019', '15/06/2019', '6/04/2019',
       '18/05/2019', '27/06/2019', '21/05/2019', '06/03/2019',
       '3/06/2019', '15/03/2019', '3/05/2019', '9/03/2019', '6/06/2019',
       '24/05/2019', '09/03/2019', '1/04/2019', '21/04/2019',
       '21/06/2019', '27/03/2019', '18/03/2019', '12/04/2019',
       '9/04/2019', '1/03/2019', '03/03/2019', '27/04/2019'], dtype=object)

In [13]:
flight[flight.Date_of_Journey.str.contains('[^0-9/]')]

Empty DataFrame
Columns: [Airline, Date_of_Journey, Source, Destination, Route, Dep_Time, Arrival_Time, Duration, Total_Stops, Additional_Info, Price]
Index: []

In [14]:
(
    pd.to_datetime
     (flight.Date_of_Journey,dayfirst=True)
)

0       2019-03-24
1       2019-05-01
2       2019-06-09
3       2019-05-12
4       2019-03-01
           ...    
10678   2019-04-09
10679   2019-04-27
10680   2019-04-27
10681   2019-03-01
10682   2019-05-09
Name: Date_of_Journey, Length: 10683, dtype: datetime64[ns]

3.3 **Source**

In [15]:
flight.Source

0        Banglore
1         Kolkata
2           Delhi
3         Kolkata
4        Banglore
           ...   
10678     Kolkata
10679     Kolkata
10680    Banglore
10681    Banglore
10682       Delhi
Name: Source, Length: 10683, dtype: object

In [16]:
flight.Source.unique()

array(['Banglore', 'Kolkata', 'Delhi', 'Chennai', 'Mumbai'], dtype=object)

3.4.**Destination**

In [17]:
flight.Destination

0        New Delhi
1         Banglore
2           Cochin
3         Banglore
4        New Delhi
           ...    
10678     Banglore
10679     Banglore
10680        Delhi
10681    New Delhi
10682       Cochin
Name: Destination, Length: 10683, dtype: object

In [18]:
flight.Destination.unique()

array(['New Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Delhi', 'Hyderabad'],
      dtype=object)

In [19]:
(

  flight.
  Destination
  .str.replace("New Delhi","Delhi")  # replace New Delhi with  Delhi
  .str.strip()
  .unique()

)

array(['Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Hyderabad'],
      dtype=object)

- New Delhi and Delhi is same as there is a single Airport so we will replace with one

3.5 **Dep_Time**

In [20]:
flight.Dep_Time

0        22:20
1        05:50
2        09:25
3        18:05
4        16:50
         ...  
10678    19:55
10679    20:45
10680    08:20
10681    11:30
10682    10:55
Name: Dep_Time, Length: 10683, dtype: object

In [21]:
(
    flight.
    Dep_Time.
    loc[lambda df_ : df_.str.contains("[^0-9:]")]

)

Series([], Name: Dep_Time, dtype: object)

In [22]:
pd.to_datetime(flight.Dep_Time,dayfirst=True, format="mixed").dt.time

0        22:20:00
1        05:50:00
2        09:25:00
3        18:05:00
4        16:50:00
           ...   
10678    19:55:00
10679    20:45:00
10680    08:20:00
10681    11:30:00
10682    10:55:00
Name: Dep_Time, Length: 10683, dtype: object

3.6 **Arrival_Time**

In [23]:
flight.Arrival_Time

0        01:10 22 Mar
1               13:15
2        04:25 10 Jun
3               23:30
4               21:35
             ...     
10678           22:25
10679           23:20
10680           11:20
10681           14:10
10682           19:15
Name: Arrival_Time, Length: 10683, dtype: object

In [24]:
(
    pd.to_datetime(flight.
    Arrival_Time,format="mixed").dt.time

)

0        01:10:00
1        13:15:00
2        04:25:00
3        23:30:00
4        21:35:00
           ...   
10678    22:25:00
10679    23:20:00
10680    11:20:00
10681    14:10:00
10682    19:15:00
Name: Arrival_Time, Length: 10683, dtype: object

3.7 **Duration**

In [25]:
flight.Duration

0        2h 50m
1        7h 25m
2           19h
3        5h 25m
4        4h 45m
          ...  
10678    2h 30m
10679    2h 35m
10680        3h
10681    2h 40m
10682    8h 20m
Name: Duration, Length: 10683, dtype: object

In [26]:
flight[~flight.Duration.str.contains('h')]

Airline Date_of_Journey  Source Destination                  Route  \
6474  Air India       6/03/2019  Mumbai   Hyderabad  BOM → GOI → PNQ → HYD   

     Dep_Time Arrival_Time Duration Total_Stops Additional_Info  Price  
6474    16:50        16:55       5m     2 stops         No info  17327

In [27]:
(
    flight.Duration.str.split(" ", expand=True)
    .set_axis(["hour", "minutes"], axis=1)
    .assign(
        hour=lambda ser_: (
            ser_
            .hour
            .str.replace("h", "")
            .str.replace("m", "")
            .astype(int)
            .mul(60)
        ),
        minutes=lambda ser_: (
            ser_
            .minutes
            .str.replace("m", "")
            .fillna("0")
            .astype(int)
        )
    )
    .sum(axis=1)
    .astype(int)
)




0         170
1         445
2        1140
3         325
4         285
         ... 
10678     150
10679     155
10680     180
10681     160
10682     500
Length: 10683, dtype: int64

3.8 **Total_Stops**

In [28]:
flight.Total_Stops.unique()

array(['non-stop', '2 stops', '1 stop', '3 stops', nan, '4 stops'],
      dtype=object)

In [29]:
(
  flight
  .Total_Stops
  .str.replace("non-stop","0")
  .str.replace('stops?',"",regex=True)
  .str.strip()
  .unique()
)

array(['0', '2', '1', '3', nan, '4'], dtype=object)

3.9 **Additional_Info**

In [30]:
(
    flight
    .Additional_Info
    .str.lower()
    .str.strip()
    .unique()

)

array(['no info', 'in-flight meal not included',
       'no check-in baggage included', '1 short layover',
       '1 long layover', 'change airports', 'business class',
       'red-eye flight', '2 long layover'], dtype=object)

------------

###**Transform and Clean Your Flight Data: A Comprehensive Dataframe Enhancement Function**

----------
### **5. Cleaning Operation**

In [31]:
def convert_2_mins(ser):
    return (
        ser
        # Split the duration string into hours and minutes based on space separator
        .str.split(" ", expand=True)

        # Rename the columns resulting from the split
        .set_axis(["hour", "minutes"], axis=1)

        # Perform the following transformations:
        .assign(
            # Convert 'hour' to minutes
            hour=lambda ser_: (
                ser_
                .hour
                # Remove 'h' from the hour values
                .str.replace("h", "")
                # Remove 'm' if present (though it should not be in 'hour')
                .str.replace("m", "")
                # Convert to integer
                .astype(int)
                # Convert hours to minutes
                .mul(60)
            ),
            # Convert 'minutes' to integer
            minutes=lambda ser_: (
                ser_
                .minutes
                # Remove 'm' from the minutes values
                .str.replace("m", "")
                # Fill NaN values with "0" (in case some durations are missing minutes)
                .fillna("0")
                # Convert to integer
                .astype(int)
            )
        )
        # Sum the 'hour' and 'minutes' columns to get the total duration in minutes
        .sum(axis=1)
        # Ensure the final result is in integer format
        .astype(int)
    )

In [32]:
# Function to clean the DataFrame
def cleaned_df(df):
    return (
        df
        # Rename columns to lowercase
        .rename(columns={i: i.lower() for i in df.columns})

        # Drop duplicate rows
        .drop_duplicates()

        # Apply various transformations to the DataFrame
        .assign(
            # Clean up 'airline' column: remove specific keywords, strip whitespace, and capitalize the names
            airline=lambda x: (
                x.airline.str.replace(' Premium economy', "")
                .str.replace(" Business", "")
                .str.strip()
                .str.title()
            ),
            # Convert 'date_of_journey' column to datetime, assuming day-first format
            date_of_journey=lambda x: pd.to_datetime(x.date_of_journey, dayfirst=True),

            # Clean up 'source' column: strip whitespace and capitalize the names
            source=lambda x: x.source.str.strip().str.title(),

            # Clean up 'destination' column: replace 'New Delhi' with 'Delhi', then strip whitespace
            destination=lambda x: x.destination.str.replace("New Delhi", "Delhi").str.strip(),

            # Convert 'dep_time' column to datetime and extract the time portion
            dep_time=lambda x: pd.to_datetime(x.dep_time, dayfirst=True, format="mixed").dt.time,

            # Convert 'arrival_time' column to datetime and extract the time portion
            arrival_time=lambda x: pd.to_datetime(x.arrival_time, format="mixed").dt.time,

            # Convert 'duration' column to minutes using the convert_2_mins function
            duration=lambda x: x.duration.pipe(convert_2_mins),

            # Clean up 'total_stops' column: replace 'non-stop' with '0', remove the word 'stops', and strip whitespace
            total_stops=lambda x: (
                x.total_stops.str.replace("non-stop", "0")
                .str.replace('stops?', "", regex=True)
                .str.strip()
            ),

            # Clean up 'additional_info' column: convert to lowercase and strip whitespace
            additional_info=lambda x: x.additional_info.str.lower().str.strip()
        )

        # Drop the 'route' column as it may not be needed after cleaning
        .drop(columns="route")

        # Drop any rows with missing values (NaN)
        .dropna()
    )


cleaned_df(flight)


airline date_of_journey    source destination  dep_time  \
0           Indigo      2019-03-24  Banglore       Delhi  22:20:00   
1        Air India      2019-05-01   Kolkata    Banglore  05:50:00   
2      Jet Airways      2019-06-09     Delhi      Cochin  09:25:00   
3           Indigo      2019-05-12   Kolkata    Banglore  18:05:00   
4           Indigo      2019-03-01  Banglore       Delhi  16:50:00   
...            ...             ...       ...         ...       ...   
10678     Air Asia      2019-04-09   Kolkata    Banglore  19:55:00   
10679    Air India      2019-04-27   Kolkata    Banglore  20:45:00   
10680  Jet Airways      2019-04-27  Banglore       Delhi  08:20:00   
10681      Vistara      2019-03-01  Banglore       Delhi  11:30:00   
10682    Air India      2019-05-09     Delhi      Cochin  10:55:00   

      arrival_time  duration total_stops additional_info  price  
0         01:10:00       170           0         no info   3897  
1         13:15:00       445           2         no info   7662  
2         04:25:00      1140           2         no info  13882  
3         23:30:00       325           1         no info   6218  
4         21:35:00       285           1         no info  13302  
...            ...       ...         ...             ...    ...  
10678     22:25:00       150           0         no info   4107  
10679     23:20:00       155           0         no info   4145  
10680     11:20:00       180           0         no info   7229  
10681     14:10:00       160           0         no info  12648  
10682     19:15:00       500           2         no info  11753  

[10462 rows x 10 columns]

--------
###**Split the data**

In [37]:
df=cleaned_df(flight)
suffled=df.sample(frac=1, random_state=42)

In [38]:

# Calculate the number of rows in the DataFrame
num_rows = suffled.shape[0]

# Calculate the number of rows for each part based on the ratio 70:20:10
part1_rows = int(num_rows * 0.70)
part2_rows = int(num_rows * 0.20)
part3_rows = num_rows - part1_rows - part2_rows  # Remaining rows for the last part

# Split the DataFrame into three parts
train = suffled.iloc[:part1_rows]
test = suffled.iloc[part1_rows:part1_rows + part2_rows]
validation = suffled.iloc[part1_rows + part2_rows:]

# Print or use the resulting splits as needed
print(f"train(70%):{train.shape[0]}")

display(train.head())
print()
print(f"test (20%):{test.shape[0]}")
display(test.head())
print()

print(f"validation (10%):{validation.shape[0] }")
display(validation.head())

train(70%):7323


airline date_of_journey    source destination  dep_time  \
2150  Jet Airways      2019-03-06  Banglore       Delhi  08:00:00   
3784     Spicejet      2019-06-06   Kolkata    Banglore  22:20:00   
714        Indigo      2019-03-18   Kolkata    Banglore  05:30:00   
7558  Jet Airways      2019-03-24    Mumbai   Hyderabad  15:50:00   
7413     Spicejet      2019-04-27  Banglore       Delhi  09:30:00   

     arrival_time  duration total_stops              additional_info  price  
2150     08:15:00      1455           1                      no info  17996  
3784     00:40:00       140           0                      no info   3873  
714      08:20:00       170           0                      no info   4462  
7558     17:20:00        90           0  in-flight meal not included   2228  
7413     12:20:00       170           0                      no info   4991


test (20%):2092


airline date_of_journey   source destination  dep_time arrival_time  \
5195     Indigo      2019-05-15    Delhi      Cochin  05:05:00     16:10:00   
1209  Air India      2019-05-01  Kolkata    Banglore  05:50:00     20:25:00   
7851  Air India      2019-06-12    Delhi      Cochin  16:55:00     19:15:00   
8160   Air Asia      2019-06-12    Delhi      Cochin  20:45:00     07:10:00   
1514     Indigo      2019-03-03    Delhi      Cochin  04:55:00     12:00:00   

      duration total_stops additional_info  price  
5195       665           1         no info   6818  
1209       875           2         no info  10991  
7851      1580           2         no info  13381  
8160       625           1         no info   6152  
1514       425           1         no info   9585


validation (10%):1047


airline date_of_journey   source destination  dep_time  \
1207              Indigo      2019-05-21    Delhi      Cochin  16:10:00   
5025           Air India      2019-05-01  Kolkata    Banglore  09:25:00   
10381           Air Asia      2019-05-09    Delhi      Cochin  16:45:00   
6867               Goair      2019-03-27    Delhi      Cochin  10:25:00   
3631   Multiple Carriers      2019-03-21    Delhi      Cochin  11:30:00   

      arrival_time  duration total_stops additional_info  price  
1207      00:55:00       525           1         no info   6778  
5025      13:45:00      1700           2         no info  16372  
10381     07:10:00       865           1         no info   7353  
6867      19:20:00       535           1         no info   5281  
3631      21:00:00       570           1         no info  13062

###**Export the splitted dataset**

In [39]:
train.to_csv("/content/train.csv", index=False)
test.to_csv("/content/test.csv", index=False)
validation.to_csv("/content/validation.csv", index=False)